Realm/ObjectStore/shared_realm.cpp

```diff
@@ -0,0 +1,430 @@
+        auto updated_realm = shared_from_this();
+        m_config.migration_function(old_realm, updated_realm);
+    };
+
+    try {
+        // update and migrate
+        begin_transaction();
+        bool changed = ObjectStore::update_realm_with_schema(read_group(), *old_config.schema,
+                                                             version, *m_config.schema,
+                                                             migration_function);
+        commit_transaction();
+        return changed;
+    }
+    catch (...) {
+        if (is_in_transaction()) {
+            cancel_transaction();
+        }
+        m_config.schema_version = old_config.schema_version;
+        m_config.schema = std::move(m_config.schema);
+        throw;
+    }
+}
+
+static void check_read_write(Realm *realm)
+{
+    if (realm->config().read_only) {
+        throw InvalidTransactionException("Can't perform transactions on read-only Realms.");
+    }
+}
+
+void Realm::verify_thread()
+{
+    if (m_thread_id != std::this_thread::get_id()) {
+        throw IncorrectThreadException("Realm accessed from incorrect thread.");
+    }
+}
+
+void Realm::begin_transaction()
+{
+    check_read_write(this);
+    verify_thread();
+
+    if (m_in_transaction) {
+        throw InvalidTransactionException("The Realm is already in a write transaction");
+    }
+
+    // make sure we have a read transaction
+    read_group();
+
+    transaction::begin(*m_shared_group, *m_history, m_delegate.get());
+    m_in_transaction = true;
+}
+
+void Realm::commit_transaction()
+{
+    check_read_write(this);
+    verify_thread();
+
+    if (!m_in_transaction) {
+        throw InvalidTransactionException("Can't commit a non-existing write transaction");
+    }
+
+    m_in_transaction = false;
+    transaction::commit(*m_shared_group, *m_history, m_delegate.get());
+    m_notifier->notify_others();
+}
+
+void Realm::cancel_transaction()
+{
+    check_read_write(this);
+    verify_thread();
+
+    if (!m_in_transaction) {
+        throw InvalidTransactionException("Can't cancel a non-existing write transaction");
+    }
+
+    m_in_transaction = false;
+    transaction::cancel(*m_shared_group, *m_history, m_delegate.get());
+}
+
+void Realm::invalidate()
+{
+    verify_thread();
+    check_read_write(this);
+
+    if (m_in_transaction) {
+        cancel_transaction();
+    }
+    if (!m_group) {
+        return;
+    }
+
+    m_shared_group->end_read();
+    m_group = nullptr;
+}
+
+bool Realm::compact()
+{
+    verify_thread();
+
+    bool success = false;
+    if (m_in_transaction) {
+        throw InvalidTransactionException("Can't compact a Realm within a write transaction");
+    }
+
+    for (auto &object_schema : *m_config.schema) {
+        ObjectStore::table_for_object_type(read_group(), object_schema.name)->optimize();
+    }
+
+    m_shared_group->end_read();
+    success = m_shared_group->compact();
+    m_shared_group->begin_read();
```

Comment:
There seems to be an implicit assumption here that `m_group` is non-null when this method is called. In the case where it's null, I think the following sequence of events will happen:
  1. `Realm::compact` begins a read transaction.
  2. The next call to `Realm::read_group` will notice that `m_group` is null and call through to `SharedGroup::begin_read()`.
  3. `SharedGroup::begin_read()` will throw since one of its preconditions is that it not be called when in a read transaction.

If we don't think it's possible to get here with a null `m_group` so documenting the assumption with an assertion is probably sufficient.